File: MiniDOMBase/DMD/Anagrafica/Persone/FindHandlers/CRegisteredFindHandler.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases
Imports minidom.XML

Partial Public Class Anagrafica


    <Serializable>
    Public Class CRegisteredFindHandler
        Inherits DBObject
        Implements IComparable, IComparable(Of CRegisteredFindHandler)


        Private m_HandlerClass As String
        Private m_EditorClass As String
        Private m_Context As String
        Private m_Priority As Integer

        Public Sub New()
            Me.m_HandlerClass = ""
            Me.m_EditorClass = ""
            Me.m_Context = ""
            Me.m_Priority = 0
        End Sub

        Public Property HandlerClass As String
            Get
                Return Me.m_HandlerClass
            End Get
            Set(value As String)
                value = Strings.Trim(value)
                Dim oldValue As String = Me.m_HandlerClass
                If (value = oldValue) Then Return
                Me.m_HandlerClass = value
                Me.DoChanged("HandlerClass", value, oldValue)
            End Set
        End Property

        Public Property EditorClass As String
            Get
                Return Me.m_EditorClass
            End Get
            Set(value As String)
                value = Strings.Trim(value)
                Dim oldValue As String = Me.m_EditorClass
                If (value = oldValue) Then Return
                Me.m_EditorClass = value
                Me.DoChanged("EditorClass", value, oldValue)
            End Set
        End Property

        Public Property Context As String
            Get
                Return Me.m_Context
            End Get
            Set(value As String)
                Dim oldValue As String = Me.m_Context
                value = Strings.Trim(value)
                If (oldValue = value) Then Return
                Me.m_Context = value
                Me.DoChanged("Context", value, oldValue)
            End Set
        End Property

        Public Property Priority As Integer
            Get
                Return Me.m_Priority
            End Get
            Set(value As Integer)
                Dim oldValue As Integer = Me.m_Priority
                If (oldValue = value) Then Return
                Me.m_Priority = value
                Me.DoChanged("Priority", value, oldValue)
            End Set
        End Property

        Public Function CompareTo(ByVal other As CRegisteredFindHandler) As Integer Implements IComparable(Of CRegisteredFindHandler).CompareTo
            Dim ret As Integer = 0
            If (ret = 0) Then ret = Me.m_Priority.CompareTo(other.m_Priority)
            If (ret = 0) Then ret = Me.m_HandlerClass.CompareTo(other.m_HandlerClass)
            Return ret
        End Function

        Private Function CompareTo(ByVal other As Object) As Integer Implements IComparable.CompareTo
            Return Me.CompareTo(CType(other, CRegisteredFindHandler))
        End Function

        Public Overrides Function ToString() As String
            Dim sb As New System.Text.StringBuilder()
            sb.Append(Me.m_HandlerClass)
            Return sb.ToString()
        End Function

        Public NotOverridable Overrides Function Equals(obj As Object) As Boolean
            Return (TypeOf (obj) Is CRegisteredFindHandler) AndAlso Me.Equals(DirectCast(obj, CRegisteredFindHandler))
        End Function

        Public Overridable Overloads Function Equals(ByVal obj As CRegisteredFindHandler) As Boolean
            Return Me.m_Context = obj.m_Context AndAlso
                   Me.m_HandlerClass = obj.m_HandlerClass AndAlso
                   Me.m_Priority = obj.m_Priority
        End Function

        Public Overrides Function GetHashCode() As Integer
            Return Me.m_HandlerClass.GetHashCode()
        End Function


        Public Overrides Function GetModule() As CModule
            Return Nothing
        End Function

        Public Overrides Function GetTableName() As String
            Return "tbl_CRMRegFindHandler"
        End Function

        Protected Friend Overrides Function GetConnection() As CDBConnection
            Return Anagrafica.Persone.GetConnection()
        End Function

        Protected Overrides Sub XMLSerialize(writer As XMLWriter)
            writer.WriteAttribute("HandlerClass", Me.m_HandlerClass)
            writer.WriteAttribute("EditorClass", Me.m_EditorClass)
            writer.WriteAttribute("Context", Me.m_Context)
            writer.WriteAttribute("Priority", Me.m_Priority)
            MyBase.XMLSerialize(writer)
        End Sub

        Protected Overrides Sub SetFieldInternal(fieldName As String, fieldValue As Object)
            Select Case fieldName
                Case "HandlerClass" : Me.m_HandlerClass = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "EditorClass" : Me.m_EditorClass = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Context" : Me.m_Context = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Priority" : Me.m_Priority = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case Else : MyBase.SetFieldInternal(fieldName, fieldValue)
            End Select

        End Sub

        Protected Overrides Function DropFromDatabase(dbConn As CDBConnection, force As Boolean) As Boolean
            Return MyBase.DropFromDatabase(dbConn, force)
        End Function

        Protected Overrides Function LoadFromRecordset(reader As DBReader) As Boolean
            Me.m_HandlerClass = reader.Read("HandlerClass", Me.m_HandlerClass)
            Me.m_EditorClass = reader.Read("EditorClass", Me.m_HandlerClass)
            Me.m_Context = reader.Read("Context", Me.m_Context)
            Me.m_Priority = reader.Read("Priority", Me.m_Priority)
            Return MyBase.LoadFromRecordset(reader)
        End Function


        Protected Overrides Function SaveToRecordset(writer As DBWriter) As Boolean
            writer.Write("HandlerClass", Me.m_HandlerClass)
            writer.Write("EditorClass", Me.m_EditorClass)
            writer.Write("Context", Me.m_Context)
            writer.Write("Priority", Me.m_Priority)
            Return MyBase.SaveToRecordset(writer)
        End Function
    End Class




End Class

File: MiniDOMBase/DMD/Anagrafica/Persone/FindHandlers/CRegisteredFindHandler.cs
using System;
using DMD;
using DMD.Databases;
using DMD.Databases.Collections;
using DMD.XML;
using static minidom.Sistema;
using minidom.repositories;

namespace minidom
{
    public partial class Anagrafica
    {

        /// <summary>
        /// Gestori di ricerca nel CRM
        /// </summary>
        [Serializable]
        public class CRegisteredFindHandler 
            : Databases.DBObject, IComparable, IComparable<CRegisteredFindHandler>
        {
            private string m_HandlerClass;
            private string m_EditorClass;
            private string m_Context;
            private int m_Priority;
            


            /// <summary>
            /// Costruttore
            /// </summary>
            public CRegisteredFindHandler()
            {
                this.m_HandlerClass = "";
                this.m_EditorClass = "";
                this.m_Context = "";
                this.m_Priority = 0;
                 
            }

            /// <summary>
            /// Nome della classe istanziata per gestire la ricerca
            /// </summary>
            public string HandlerClass
            {
                get
                {
                    return m_HandlerClass;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_HandlerClass;
                    if ((value ?? "") == (oldValue ?? ""))
                        return;
                    m_HandlerClass = value;
                    DoChanged("HandlerClass", value, oldValue);
                }
            }
 
            /// <summary>
            /// Nome della classe dell'editor
            /// </summary>
            public string EditorClass
            {
                get
                {
                    return m_EditorClass;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_EditorClass;
                    if ((value ?? "") == (oldValue ?? ""))
                        return;
                    m_EditorClass = value;
                    DoChanged("EditorClass", value, oldValue);
                }
            }

            /// <summary>
            /// Contesto in cui è valido l'handler
            /// </summary>
            public string Context
            {
                get
                {
                    return m_Context;
                }

                set
                {
                    string oldValue = m_Context;
                    value = DMD.Strings.Trim(value);
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_Context = value;
                    DoChanged("Context", value, oldValue);
                }
            }

            /// <summary>
            /// Priorietà dell'handler
            /// </summary>
            public int Priority
            {
                get
                {
                    return m_Priority;
                }

                set
                {
                    int oldValue = m_Priority;
                    if (oldValue == value)
                        return;
                    m_Priority = value;
                    DoChanged("Priority", value, oldValue);
                }
            }

            /// <summary>
            /// Compara gli handler per ordinarli in funzione della priorità
            /// </summary>
            /// <param name="other"></param>
            /// <returns></returns>
            public int CompareTo(CRegisteredFindHandler other)
            {
                int ret = 0;
                if (ret == 0) ret = m_Priority.CompareTo(other.m_Priority);
                if (ret == 0) ret = m_HandlerClass.CompareTo(other.m_HandlerClass);
                return ret;
            }

            int IComparable.CompareTo(object other)
            {
                return CompareTo((CRegisteredFindHandler)other);
            }

            /// <summary>
            /// Restituisce una stringa che rappresenta l'oggetto
            /// </summary>
            /// <returns></returns>
            public override string ToString()
            {
                var sb = new System.Text.StringBuilder();
                sb.Append(m_HandlerClass);
                return sb.ToString();
            }

            /// <summary>
            /// Restituisce true se i due oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public override bool Equals(Databases.DBObject obj)
            {
                return (obj is CRegisteredFindHandler) && this.Equals((CRegisteredFindHandler)obj);
            }

            /// <summary>
            /// Restituisce true se i due oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public virtual bool Equals(CRegisteredFindHandler obj)
            {
                return DMD.Strings.EQ(this.m_Context , obj.m_Context)
                     && DMD.Strings.EQ(this.m_HandlerClass, obj.m_HandlerClass )
                     && DMD.Integers.EQ(this.m_Priority, obj.m_Priority);
            }

            /// <summary>
            /// Restituisce il codice hash dell'oggetto
            /// </summary>
            /// <returns></returns>
            public override int GetHashCode()
            {
                return HashCalculator.Calculate(this.m_HandlerClass, this.m_EditorClass, this.m_Context);
            }

            /// <summary>
            /// Restituisce un riferimento al repository
            /// </summary>
            /// <returns></returns>
            public override CModulesClass GetModule()
            {
                return minidom.Anagrafica.RegisteredFindPersonaHandlers;
            }

            /// <summary>
            /// Restituisce il nome della tabella
            /// </summary>
            /// <returns></returns>
            public override string GetTableName()
            {
                return "tbl_CRMRegFindHandler";
            }

        
            /// <summary>
            /// Serializzazione XML
            /// </summary>
            /// <param name="writer"></param>
            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("HandlerClass", m_HandlerClass);
                writer.WriteAttribute("EditorClass", m_EditorClass);
                writer.WriteAttribute("Context", m_Context);
                writer.WriteAttribute("Priority", m_Priority);
                base.XMLSerialize(writer);
                 
            }

            /// <summary>
            /// Deserializzazione XML
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "HandlerClass":
                        {
                            m_HandlerClass = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "EditorClass":
                        {
                            m_EditorClass = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Context":
                        {
                            m_Context = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Priority":
                        {
                            m_Priority = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }
                     

                    default:
                        {
                            base.SetFieldInternal(fieldName, fieldValue);
                            break;
                        }
                }
            }

           

            /// <summary>
            /// Carica dal DB
            /// </summary>
            /// <param name="reader"></param>
            /// <returns></returns>
            protected override bool LoadFromRecordset(DBReader reader)
            {
                m_HandlerClass = reader.Read("HandlerClass", m_HandlerClass);
                m_EditorClass = reader.Read("EditorClass", m_HandlerClass);
                m_Context = reader.Read("Context", m_Context);
                m_Priority = reader.Read("Priority", m_Priority);
                
                return base.LoadFromRecordset(reader);
            }

            /// <summary>
            /// Salva nel DB
            /// </summary>
            /// <param name="writer"></param>
            /// <returns></returns>
            protected override bool SaveToRecordset(DBWriter writer)
            {
                writer.Write("HandlerClass", m_HandlerClass);
                writer.Write("EditorClass", m_EditorClass);
                writer.Write("Context", m_Context);
                writer.Write("Priority", m_Priority);
                return base.SaveToRecordset(writer);
            }

            /// <summary>
            /// Prepara lo schema per l'oggetto
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaFields(DBTable table)
            {
                base.PrepareDBSchemaFields(table);

                var c = table.Fields.Ensure("HandlerClass", typeof(string), 255);
                c = table.Fields.Ensure("EditorClass", typeof(string), 255);
                c = table.Fields.Ensure("Context", typeof(string), 255);
                c = table.Fields.Ensure("Priority", typeof(int), 0);
            }

            /// <summary>
            /// Prepara gli indici
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaConstraints(DBTable table)
            {
                base.PrepareDBSchemaConstraints(table);

                var c = table.Constraints.Ensure("idxClass", new string[] {"HandlerClass", "EditorClass", "Context", "Priority" }, DBFieldConstraintFlags.None);

            }
        }
    }
}